internal/controller/resilient_deployment_workflow.go

```diff
@@ -0,0 +1,625 @@
+
+// WARNING: level varible is only here for the basic dev work and should not end up in the finished feature
+// FIXME: don't merge to main with value of zero, set to one.
+var level = 1
+
+const (
+	ExperimentalResilienceFeature = "ExperimentalResilienceFeature"
+	ResilienceFeatureAnnotation   = "kuadrant.io/experimental-dont-use-resilient-data-plane"
+	LimitadorReplicas             = 2
+	LimitadorPDB                  = 1
+
+	Resource10Mi = "10Mi"
+	Resource10m  = "10m"
+)
+
+func NewResilienceDeploymentWorkflow(client *dynamic.DynamicClient) *controller.Workflow {
+	return &controller.Workflow{
+		Precondition:  NewResilienceDeploymentPrecondition().Subscription().Reconcile,
+		Tasks:         NewResilienceDeploymentTasks(client),
+		Postcondition: NewResilienceDeploymentPostcondition().Subscription().Reconcile,
+	}
+}
+
+// INFO: Precontion Section
+
+func NewResilienceDeploymentPrecondition() *ResilienceDeploymentPrecondition {
+	return &ResilienceDeploymentPrecondition{}
+}
+
+type ResilienceDeploymentPrecondition struct{}
+
+func (r *ResilienceDeploymentPrecondition) Subscription() controller.Subscription {
+	return controller.Subscription{
+		ReconcileFunc: r.run,
+		Events: []controller.ResourceEventMatcher{
+			{Kind: &kuadrantv1beta1.KuadrantGroupKind},
+			// TODO: review the limitador event watch after the feature comes out of experimental
+			{Kind: &kuadrantv1beta1.LimitadorGroupKind},
+		},
+	}
+}
+
+func (r *ResilienceDeploymentPrecondition) run(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error, state *sync.Map) error {
+	logger := controller.LoggerFromContext(ctx).WithName("ResilienceDeploymentPrecondition")
+	logger.V(level).Info("ResilienceDeployment Precondition", "status", "started")
+	defer logger.V(level).Info("ResilienceDeployment Precondition", "status", "completed")
+
+	state.Store(ExperimentalResilienceFeature, isExperimentalFeatureEnabled(topology))
+	state.Store("resilienceError", false)
+
+	return nil
+}
+
+// INFO: Task Section
+
+func NewResilienceDeploymentTasks(client *dynamic.DynamicClient) []controller.ReconcileFunc {
+	return []controller.ReconcileFunc{
+		NewResilienceAuthorizationReconciler().Subscription().Reconcile,
+		NewResilienceCounterStorageReconciler(client).Subscription().Reconcile,
+		NewResilienceRateLimitingReconciler(client).Subscription().Reconcile,
+	}
+}
+
+func NewResilienceAuthorizationReconciler() *ResilienceAuthorizationReconciler {
+	return &ResilienceAuthorizationReconciler{}
+}
+
+type ResilienceAuthorizationReconciler struct{}
+
+func (r *ResilienceAuthorizationReconciler) Subscription() controller.Subscription {
+	return controller.Subscription{
+		ReconcileFunc: r.reconcile,
+		Events: []controller.ResourceEventMatcher{
+			{Kind: &kuadrantv1beta1.KuadrantGroupKind},
+		},
+	}
+}
+
+func (r *ResilienceAuthorizationReconciler) reconcile(ctx context.Context, _ []controller.ResourceEvent, _ *machinery.Topology, _ error, state *sync.Map) error {
```

Comment:
one thought I had here was the resourceEvents contain the old and new object so perhaps these could be used to compare what has changed. Perhaps not needed 🤷

Reply:
I think it will give you the old and new version of the object. This however can not be used as the old version that you will be give is the old version on the cluster, and the last version you successfully reconciled. If you could compare the new version with the past reconciled version, then yes we could have used that.

Reply:
Ok so what exactly does it give you then?